polar-core/src/partial/constraints.rs

```diff
@@ -0,0 +1,501 @@
+                | Operator::Eq
+                | Operator::Neq
+        ));
+
+        let op = Operation {
+            operator,
+            args: vec![self.variable_term(), other],
+        };
+
+        self.operations.push(op);
+    }
+
+    /// Add lookup of `field` assigned to `value` on `self.
+    ///
+    /// Returns: A partial expression for `value`.
+    pub fn lookup(&mut self, field: Term, value: Term) -> Term {
+        // Note this is a 2-arg lookup (Dot) not 3-arg. (Pre rewrite).
+        assert!(matches!(field.value(), Value::String(_)));
+
+        self.operations.push(op!(
+            Unify,
+            value.clone(),
+            term!(op!(Dot, self.variable_term(), field))
+        ));
+
+        let name = value.value().as_symbol().unwrap();
+        Term::new_temporary(Value::Partial(Constraints::new(name.clone())))
+    }
+
+    pub fn into_term(self) -> Term {
+        Term::new_temporary(Value::Partial(self))
+    }
+
+    /// Return the expression represented by this partial's constraints.
+    pub fn into_expression(self) -> Term {
+        Term::new_temporary(Value::Expression(Operation {
+            operator: Operator::And,
+            args: self
+                .operations
+                .into_iter()
+                .map(|op| Term::new_temporary(Value::Expression(op)))
+                .collect(),
+        }))
+    }
+
+    pub fn clone_with_name(&self, name: Symbol) -> Self {
+        let mut new = self.clone();
+        new.variable = name;
+        new
+    }
+
+    pub fn clone_with_operations(&self, operations: Vec<Operation>) -> Self {
+        let mut new = self.clone();
+        new.operations = operations;
+        new
+    }
+
+    pub fn name(&self) -> &Symbol {
+        &self.variable
+    }
+
+    fn variable_term(&self) -> Term {
+        Term::new_temporary(Value::Variable(sym!("_this")))
+    }
+}
+
+#[derive(Clone)]
+struct IsaConstraintCheck {
+    existing: Vec<Operation>,
+    proposed_tag: Option<Symbol>,
+    result: Option<bool>,
+    last_call_id: u64,
+}
+
+impl IsaConstraintCheck {
+    pub fn new(existing: Vec<Operation>, mut proposed: Operation) -> Self {
+        let right = proposed.args.pop().unwrap();
+        let proposed_tag = if let Value::Pattern(Pattern::Instance(instance)) = right.value() {
+            Some(instance.tag.clone())
+        } else {
+            None
+        };
+
+        Self {
+            existing,
+            proposed_tag,
+            result: None,
+            last_call_id: 0,
+        }
+    }
+
+    /// Check if constraint is compatible with proposed.
+    ///
+    /// Returns: None if compatible, QueryEvent::Done { false } if incompatible,
+    /// or QueryEvent to ask for compatibility.
+    fn check_constraint(
+        &mut self,
+        mut constraint: Operation,
+        counter: &Counter,
+    ) -> Option<QueryEvent> {
+        if constraint.operator != Operator::Isa {
+            return None;
+        }
```

Comment:
Yes, it could.  But, these constraints are approximate.  It's okay if conflicting constraints are pushed.  It is incorrect if constraints that don't conflict are marked as conflicting.